waltz-ng/client/logical-flow/components/bulk-logical-flow-parser/bulk-logical-flow-parser.js

```diff
@@ -0,0 +1,201 @@
+                });
+        })
+}
+
+
+function controller($q, serviceBroker) {
+    const vm = initialiseData(this, initialState);
+
+    const loadIdentifierToEntityRefMap = async () => {
+        return await serviceBroker
+            .loadViewData(CORE_API.ApplicationStore.findAll)
+            .then(r => _.keyBy(r.data, k => _.toLower(k.assetCode)));
+    };
+
+    const loadCodeToDataTypeMap = async () => {
+        return await serviceBroker
+            .loadViewData(CORE_API.DataTypeStore.findAll)
+            .then(r => _.keyBy(r.data, k => _.toLower(k.code)));
+    };
+
+    const isComplete = () => {
+        const allParsedRefs = _.flatMap(vm.parsedData, d => [d.source, d.target, d.dataType]);
+        return _.every(allParsedRefs, p => p.entityRef != null);
+    };
+
+    const parseErrorCount = (data = []) => {
+        const allParsedRefs = _.flatMap(data, d => [d.source, d.target, d.dataType]);
+        return _.sumBy(allParsedRefs, p => p.entityRef == null ? 1 : 0);
+    };
+
+
+    vm.$onInit = async () => {
+        vm.loading = true;
+        const [entityRefsByAssetCode, dataTypesByCode] = await $q.all([
+            loadIdentifierToEntityRefMap(),
+            loadCodeToDataTypeMap()
+        ]);
+
+        vm.columnResolvers = {
+            source: (identifier) => resolveEntityRef(entityRefsByAssetCode, 'APPLICATION', identifier),
+            target: (identifier) => resolveEntityRef(entityRefsByAssetCode, 'APPLICATION', identifier),
+            dataType: (identifier) => resolveEntityRef(dataTypesByCode, 'DATA_TYPE', identifier),
+        }
+    };
+
+
+    vm.$onChanges = async (changes) => {
+        if(vm.columnMappings && vm.sourceData) {
+            const mappedData = mapColumns(vm.columnMappings, vm.sourceData, vm.columnResolvers);
+
+            if(mappedData.length > 0 && parseErrorCount(mappedData) == 0) {
+                //compare against logical flows in database
+
+                const sourcesAndTargets = _.map(mappedData, p => ({source: p.source.entityRef, target: p.target.entityRef}));
+                const existingFlows = await findExistingLogicalFlowsAndDecorators(serviceBroker, sourcesAndTargets);
+
+                // nesting: source -> target -> data type
+                const existingFlowsNested = nest()
+                    .key(flow => refToString(flow.source))
+                    .key(flow => refToString(flow.target))
+                    .key(flow => refToString(flow.decorator.decoratorEntity))
+                    .object(existingFlows);
+
+                const parsedWithExisting = _.map(mappedData, p => {
+                    const sourceRefString = refToString(p.source.entityRef);
+                    const targetRefString = refToString(p.target.entityRef);
+                    const dataTypeString = refToString(p.dataType.entityRef);
+                    const existingFlows = _.get(existingFlowsNested, `[${sourceRefString}][${targetRefString}][]${dataTypeString}`);
```

Comment:
Not sure if I'm impressed or worried ;)

Reply:
Still can't quite work out the last bit (is it supposed to be inside the square brackets....)

Reply:
agreed, typo, fixed